### PR TITLE
Replace timezone shapefile parsing with geojson parsing

### PR DESCRIPTION
**Pull request by karenzshea**
# Issue

Closes https://github.com/Project-OSRM/osrm-backend/issues/3967

Shapefile dependency is finicky and possessive. This implements timezone parsing with rapidjson instead! Thus expecting timezone files to be provided as geojson files...

## Tasklist
 - [x] Remove code relating to shapefile that can be removed
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [x] add unit tests for geojson parsing
 - [x] review
 - [ ] adjust for comments

## Requirements / Relations
5.8

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Looks like you're bundling rapidjson in the repo. now. A few questions wrt. that:

Is rapidjson packaged in Linux distributions we care about and can we use that?

If not, is it worth it to bundle rapidjson just for parsing geojson shapefiles? Or can we convert the shapefile polygones (?) into e.g. WKT we can then read with boost geometry's [`read_wkt`](http://www.boost.org/doc/libs/1_64_0/libs/geometry/doc/html/geometry/reference/io/wkt/read_wkt.html)?

File: src/util/timezones.cpp [hunk 1]
```diff
@@ -16,6 +16,8 @@
 #include <string>
 #include <unordered_map>
 
+#include <time.h>
```
Comment (karenzshea) [inline src/util/timezones.cpp]: should we also #ifdef this line?

Comment (oxidase) [inline src/util/timezones.cpp]: it is ok both `localtime` versions are declared `time.h` in http://pubs.opengroup.org/onlinepubs/009695399/functions/localtime.html
somewhere the file was already included in non win32 builds

File: src/util/timezones.cpp [hunk 2]
```diff
@@ -81,50 +88,75 @@ void Timezoner::LoadLocalTimesRTree(const std::string &tz_shapes_filename, std::
...
             polygons.emplace_back(boost::geometry::return_envelope<box_t>(polygon),
                                   local_times.size());
 
             // Get time zone name and emplace polygon and local time for the UTC input
-            const auto tzname = DBFReadStringAttribute(dbfhandle, shape, tzid);
-            local_times.emplace_back(local_time_t{polygon, get_local_time_in_tz(tzname)});
-
-            // std::cout << boost::geometry::dsv(boost::geometry::return_envelope<box_t>(polygon))
-            //           << " " << tzname << " " << asctime(&local_times.back().second);
+            const auto tzname =
```
Comment (oxidase) [inline src/util/timezones.cpp]: const auto &

File: CMakeLists.txt [hunk 3]
```diff
@@ -447,11 +447,6 @@ if(ENABLE_MASON)
   add_dependency_includes(${MASON_PACKAGE_tbb_INCLUDE_DIRS})
   set(TBB_LIBRARIES ${MASON_PACKAGE_tbb_LDFLAGS})
 
-  mason_use(libshp2 VERSION ${MASON_LIBSHP_VERSION})
```
Comment (TheMarex) [inline CMakeLists.txt]: Yay 🎉 death to all shapefiles.

File: CMakeLists.txt [hunk 4]
```diff
@@ -698,10 +685,11 @@ install(TARGETS osrm-components DESTINATION bin)
 if(BUILD_TOOLS)
   message(STATUS "Activating OSRM internal tools")
   add_executable(osrm-io-benchmark src/tools/io-benchmark.cpp $<TARGET_OBJECTS:UTIL>)
-  target_link_libraries(osrm-io-benchmark ${BOOST_BASE_LIBRARIES} ${MAYBE_SHAPEFILE})
+  target_link_libraries(osrm-io-benchmark ${BOOST_BASE_LIBRARIES})
 
   install(TARGETS osrm-io-benchmark DESTINATION bin)
 
+  find_package(Shapefile)
```
Comment (TheMarex) [inline CMakeLists.txt]: Could we integrate the geojson parser in the tool too?

File: include/util/geojson_validation.hpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+#include "util/log.hpp"
+
+#include "rapidjson/document.h"
+
+namespace osrm
+{
+namespace util
+{
+
+inline void ValidateCoordinate(const rapidjson::Value &coordinate)
```
Comment (TheMarex) [inline include/util/geojson_validation.hpp]: free functions should start with a small first letter: `validateCoordinate`.

File: include/util/geojson_validation.hpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        for (rapidjson::SizeType i = 0; i < coordinate.Size(); i++)
+        {
+            if (!coordinate[i].IsNumber() && !coordinate[i].IsDouble())
+                throw osrm::util::exception("Feature geometry has a non-number coordinate.");
+        }
+    }
+}
+
+inline void ValidateFeature(const rapidjson::Value &feature)
```
Comment (TheMarex) [inline include/util/geojson_validation.hpp]: Nice! Resilient code. Function name should be `validateFeature` as above.

**Comment by TheMarex** [comment]
As per chat with @karenzshea: Fixed the remaining PR issue to get this merged in time for the pre-release of 5.8.0.

**Pull request merged**